yt_dlp/extractor/niconico.py

```diff
@@ -409,19 +417,91 @@ def _real_extract(self, url):
                     webpage, 'error reason', default=None)
                 if not error_msg:
                     raise
-                raise ExtractorError(re.sub(r'\s+', ' ', error_msg), expected=True)
+                raise ExtractorError(clean_html(error_msg), expected=True)
+
+        club_joined = traverse_obj(api_data, ('channel', 'viewer', 'follow', 'isFollowed', {bool}))
+        if club_joined is None:
+            fail_msg = self._html_search_regex(
+                r'<p[^>]+\bclass="fail-message"[^>]*>(?P<msg>.+)</p>',
+                webpage, 'fail message', default=None, group='msg')
+            if fail_msg:
+                self.raise_login_required(clean_html(fail_msg), metadata_available=True)
+        elif not club_joined:
+            self.raise_login_required('This video is for members only', metadata_available=True)
 
         formats = []
 
         def get_video_info(*items, get_first=True, **kwargs):
             return traverse_obj(api_data, ('video', *items), get_all=not get_first, **kwargs)
 
-        quality_info = api_data['media']['delivery']['movie']
-        session_api_data = quality_info['session']
-        for (audio_quality, video_quality, protocol) in itertools.product(quality_info['audios'], quality_info['videos'], session_api_data['protocols']):
-            fmt = self._extract_format_for_quality(video_id, audio_quality, video_quality, protocol)
-            if fmt:
-                formats.append(fmt)
+        dmc_data = traverse_obj(api_data, ('media', 'delivery', 'movie', {
+            'audios': ('audios', ..., {dict}),
+            'videos': ('videos', ..., {dict}),
+            'protocols': ('session', 'protocols', ..., {str}),
+        }))
+        if len(dmc_data) == 3:
+            for (audio_quality, video_quality, protocol) in itertools.product(
+                    dmc_data['audios'], dmc_data['videos'], dmc_data['protocols']):
+                fmt = self._extract_format_for_quality(video_id, audio_quality, video_quality, protocol)
+                if fmt:
+                    formats.append(fmt)
+
+        # Getting all audio formats results in extracting duplicate video formats.
+        # Duplicate ones will be filtered out later.
+        dms_data = traverse_obj(api_data, ({
```

Comment:
I think we should move the DMC and DMS formats extraction into methods; we don't have to indent everything under if statements and can keep the code cleaner/more readable IMO

Reply:
I pushed a refactor, let me know what you think, and if you approve, please test the changes